api/handler/acl.go

```diff
@@ -0,0 +1,1185 @@
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	"github.com/nspcc-dev/neofs-api-go/pkg/acl/eacl"
+	"github.com/nspcc-dev/neofs-api-go/pkg/object"
+	"github.com/nspcc-dev/neofs-s3-gw/api"
+	"github.com/nspcc-dev/neofs-s3-gw/api/layer"
+	"go.uber.org/zap"
+)
+
+var (
+	writeOps = []eacl.Operation{eacl.OperationPut, eacl.OperationDelete}
+	readOps  = []eacl.Operation{eacl.OperationGet, eacl.OperationHead,
+		eacl.OperationSearch, eacl.OperationRange, eacl.OperationRangeHash}
+	fullOps = []eacl.Operation{eacl.OperationGet, eacl.OperationHead, eacl.OperationPut,
+		eacl.OperationDelete, eacl.OperationSearch, eacl.OperationRange, eacl.OperationRangeHash}
+)
+
+var actionToOpMap = map[string][]eacl.Operation{
+	"s3:DeleteObject": {eacl.OperationDelete},
+	"s3:GetObject":    readOps,
+	"s3:PutObject":    {eacl.OperationPut},
+	"s3:ListBucket":   readOps,
+}
+
+const (
+	arnAwsPrefix     = "arn:aws:s3:::"
+	allUsersWildcard = "*"
+	allUsersGroup    = "http://acs.amazonaws.com/groups/global/AllUsers"
+)
+
+// AWSACL is aws permission constants.
+type AWSACL string
+
+const (
+	aclFullControl AWSACL = "FULL_CONTROL"
+	aclWrite       AWSACL = "WRITE"
+	aclRead        AWSACL = "READ"
+)
+
+// GranteeType is aws grantee permission type constants.
+type GranteeType string
+
+const (
+	acpCanonicalUser         GranteeType = "CanonicalUser"
+	acpAmazonCustomerByEmail GranteeType = "AmazonCustomerByEmail"
+	acpGroup                 GranteeType = "Group"
+)
+
+type bucketPolicy struct {
+	Version   string      `json:"Version"`
+	ID        string      `json:"Id"`
+	Statement []statement `json:"Statement"`
+}
+
+type statement struct {
+	Sid       string    `json:"Sid"`
+	Effect    string    `json:"Effect"`
+	Principal principal `json:"Principal"`
+	Action    []string  `json:"Action"`
+	Resource  []string  `json:"Resource"`
+}
+
+type principal struct {
+	AWS           string `json:"AWS,omitempty"`
+	CanonicalUser string `json:"CanonicalUser,omitempty"`
+}
+
+type ast struct {
+	Resources []*astResource
+}
+
+type astResource struct {
+	Name       string
+	Operations []*astOperation
+}
+
+type astOperation struct {
+	Users  []string
+	Role   eacl.Role
+	Op     eacl.Operation
+	Action eacl.Action
+}
+
+func (h *handler) GetBucketACLHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		req = mux.Vars(r)
+		bkt = req["bucket"]
+		rid = api.GetRequestID(r.Context())
+	)
+
+	bucketACL, err := h.obj.GetBucketACL(r.Context(), bkt)
+	if err != nil {
+		h.writeError(w, r, "could not fetch bucket acl", rid, bkt, err)
+		return
+	}
+
+	if err := checkOwner(bucketACL.Info, r.Header.Get(api.AmzExpectedBucketOwner)); err != nil {
+		h.writeError(w, r, "expected owner doesn't match", rid, bkt, err)
+		return
+	}
+
+	if err = api.EncodeToResponse(w, h.encodeBucketACL(bucketACL)); err != nil {
+		h.registerAndSendError(w, r, err, "something went wrong")
+	}
+}
+
+func (h *handler) writeError(w http.ResponseWriter, r *http.Request, msg, rid, bkt string, err error) {
```

Comment:
This function duplicates `func (h *handler) registerAndSendError(w http.ResponseWriter, r *http.Request, err error, logText string)`. 

We need the function like `writeError` everywhere in handler.  Maybe we should create a separate file (util.go, for example) and put functions like this into this file. If you don't like a name `registerAndSendError` you can use `writeError`, but could you please replace all usages of `registerAndSendError` by `WriteError`?

Reply:
Also as I see we use `rid` (at this moment) only in logs messages. Maybe we can extract `rid` from `http.Request.Context()` inside `writeError` and don't pass `rid` together with `http.Request`?

Reply:
I think we can create a separate issue for this (refactor so that the `registerAndSendError` function takes `api.ReqInfo` to log at least method and bucket name). For now, let's use the `registerAndSendError` function everywhere.